packages/monaco/src/browser/monaco-editor-model.ts

```diff
@@ -333,28 +333,76 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
-        this.onWillSaveModelEmitter.fire({
-            model: this, reason,
-            waitUntil: thenable => {
-                waitables.push(thenable);
+        type EditContributor = Thenable<monaco.editor.IIdentifiedSingleEditOperation[]>;
+
+        let didTimeout = false;
+        let timeoutHandle: number;
+
+        const shouldStop: () => boolean = () => token.isCancellationRequested || didTimeout;
+
+        // tslint:disable-next-line:no-any
+        const timeoutPromise = new Promise((resolve, reject) => timeoutHandle = <any>setTimeout(() => {
+            didTimeout = true;
+            reject(new Error('onWillSave listener loop timeout'));
+        }, 1000));
+
+        const firing = this.onWillSaveModelEmitter.sequence(async listener => {
+            if (shouldStop()) {
+                return Emitter.StopSequence;
+            }
+            const waitables: EditContributor[] = [];
+            const { version } = this;
+
+            const event = {
+                model: this, reason,
+                waitUntil: (thenable: EditContributor) => {
+                    if (Object.isFrozen(waitables)) {
+                        throw new Error('waitUntil cannot be called asynchronously.');
+                    }
+                    waitables.push(thenable);
+                }
+            };
+
+            // Fire.
+            try {
+                listener(event);
+            } catch (err) {
+                console.error(err);
+                return; // Continue with next listener.
             }
-        });
 
-        // Wait for all listeners to resolve the edits.
-        return Promise.all(waitables).then(allOperations => {
-            if (token.isCancellationRequested) {
-                return;
+            // Asynchronous calls to `waitUntil` should fail.
+            Object.freeze(waitables);
+
+            // Wait for all promises.
+            const edits = await Promise.all(waitables).then(allOperations =>
+                ([] as monaco.editor.IIdentifiedSingleEditOperation[]).concat(...allOperations)
+            );
+            if (shouldStop()) {
+                return Emitter.StopSequence;
+            }
+
+            // In a perfect world, we should only apply edits if document is clean.
+            if (version !== this.version) {
+                console.error('onWillSave listeners should provide edits, not directly alter the document.');
             }
-            for (const operations of allOperations) {
-                this.applyEdits(operations, {
+
+            // Finally apply edits provided by this listener before firing the next.
+            if (edits && edits.length > 0) {
+                this.applyEdits(edits, {
                     ignoreDirty: true,
                 });
             }
         });
+
+        return Promise.race([
+            timeoutPromise,
+            firing,
+        ]).catch(e => {
+            console.error(e);
+        }).then(() => {
```

Comment:
Want the timer cleaned only if an error happened? You can use `await` + `finally`:
```ts
        try {
            await Promise.race([timeoutPromise, firing]);
        } catch (e) {
            console.error(e);
        } finally {
            clearTimeout(timeoutHandle);
        }
```

Reply:
Just tried the following snippet:

```js
Promise.reject(new Error('meh'))
    .catch(e => console.error(e.message))
    .then(() => console.log('done'))

Promise.resolve(1)
    .catch(e => console.error(e.message))
    .then(() => console.log('done'))

// outputs:
meh
done
done
```

The last `then` statement is called every time, acting a bit like a `finally` clause.

But I feel like it looks clearer using `try/catch/finally`, so will do!